src/dataloader/mod.rs

```diff
@@ -0,0 +1,614 @@
+pub struct DataLoader<K: Send + Sync + Eq + Clone + Hash + 'static, T: Loader<K>, C = NoCache> {
+  inner: Arc<DataLoaderInner<K, T>>,
+  cache_factory: C,
+  delay: Duration,
+  max_batch_size: usize,
+  disable_cache: AtomicBool,
+  spawner: Box<dyn Fn(BoxFuture<'static, ()>) + Send + Sync>,
+}
+
+impl<K, T> DataLoader<K, T, NoCache>
+where
+  K: Send + Sync + Hash + Eq + Clone + 'static,
+  T: Loader<K>,
+{
+  /// Use `Loader` to create a [DataLoader] that does not cache records.
+  pub fn new<S, R>(loader: T, spawner: S) -> Self
+  where
+    S: Fn(BoxFuture<'static, ()>) -> R + Send + Sync + 'static,
+  {
+    Self {
+      inner: Arc::new(DataLoaderInner { requests: Mutex::new(Default::default()), loader }),
+      cache_factory: NoCache,
+      delay: Duration::from_millis(1),
+      max_batch_size: 1000,
+      disable_cache: false.into(),
+      spawner: Box::new(move |fut| {
+        spawner(fut);
+      }),
+    }
+  }
+}
+
+impl<K, T, C: CacheFactory> DataLoader<K, T, C>
+where
+  K: Send + Sync + Hash + Eq + Clone + 'static,
+  T: Loader<K>,
+{
+  /// Use `Loader` to create a [DataLoader] with a cache factory.
+  pub fn with_cache<S, R>(loader: T, spawner: S, cache_factory: C) -> Self
+  where
+    S: Fn(BoxFuture<'static, ()>) -> R + Send + Sync + 'static,
+  {
+    Self {
+      inner: Arc::new(DataLoaderInner { requests: Mutex::new(Default::default()), loader }),
+      cache_factory,
+      delay: Duration::from_millis(1),
+      max_batch_size: 1000,
+      disable_cache: false.into(),
+      spawner: Box::new(move |fut| {
+        spawner(fut);
+      }),
+    }
+  }
+
+  /// Specify the delay time for loading data, the default is `1ms`.
+  #[must_use]
+  pub fn delay(self, delay: Duration) -> Self {
+    Self { delay, ..self }
+  }
+
+  /// pub fn Specify the max batch size for loading data, the default is
+  /// `1000`.
+  ///
+  /// If the keys waiting to be loaded reach the threshold, they are loaded
+  /// immediately.
+  #[must_use]
+  pub fn max_batch_size(self, max_batch_size: usize) -> Self {
+    Self { max_batch_size, ..self }
+  }
+
+  /// Get the loader.
+  #[inline]
+  pub fn loader(&self) -> &T {
+    &self.inner.loader
+  }
+
+  /// Enable/Disable cache of all loaders.
+  pub fn enable_all_cache(&self, enable: bool) {
+    self.disable_cache.store(!enable, Ordering::SeqCst);
+  }
+
+  /// Enable/Disable cache of specified loader.
+  pub fn enable_cache(&self, enable: bool)
+  where
+    K: Send + Sync + Hash + Eq + Clone + 'static,
+    T: Loader<K>,
+  {
+    let tid = TypeId::of::<K>();
+    let mut requests = self.inner.requests.lock().unwrap();
+    let typed_requests = requests.get_mut(&tid).unwrap();
+    typed_requests.disable_cache = !enable;
+  }
+
+  /// Use this `DataLoader` load a data.
+  #[cfg_attr(feature = "tracing", instrument(skip_all))]
+  pub async fn load_one(&self, key: K) -> Result<Option<T::Value>, T::Error>
+  where
+    K: Send + Sync + Hash + Eq + Clone + 'static,
```

Comment:
instead of copying the same type conditions all over again we can use type alias